### PR TITLE
Add snapshot types

### PR DESCRIPTION
**Pull request by thibault-martinez**
## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes
- [ ] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

File: bee-ledger/src/types/snapshot/kind.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    fn pack<W: Write>(&self, writer: &mut W) -> Result<(), Self::Error> {
+        (*self as u8).pack(writer)?;
+
+        Ok(())
+    }
+
+    fn unpack_inner<R: Read + ?Sized, const CHECK: bool>(reader: &mut R) -> Result<Self, Self::Error> {
+        Ok(match u8::unpack_inner::<R, CHECK>(reader)? {
```
Comment (zesterer) [inline bee-ledger/src/types/snapshot/kind.rs:34]: I wonder whether we should seek to derive `num`'s `FromPrimitive` for these types at some point to get rid of this sort of thing

File: bee-ledger/src/types/snapshot/milestone_diff.rs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        0u32.packed_len()
+            + std::mem::size_of_val(&MilestonePayload::KIND)
+            + self.milestone.packed_len()
+            + if let Some((treasury_output, milestone_id)) = self.consumed_treasury.as_ref() {
+                milestone_id.packed_len() + treasury_output.packed_len()
+            } else {
+                0
+            }
+            + 0u64.packed_len()
+            + self.created_outputs.iter().fold(0, |acc, (output_id, created_output)| {
```
Comment (zesterer) [inline bee-ledger/src/types/snapshot/milestone_diff.rs]: `map` + `sum` instead of `fold`?

File: bee-ledger/src/types/snapshot/milestone_diff.rs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    /// Returns the consumed outputs of a `MilestoneDiff`.
+    pub fn consumed(&self) -> &HashMap<OutputId, (CreatedOutput, ConsumedOutput)> {
+        &self.consumed_outputs
+    }
+}
+
+impl Packable for MilestoneDiff {
+    type Error = Error;
+
+    fn packed_len(&self) -> usize {
```
Comment (zesterer) [inline bee-ledger/src/types/snapshot/milestone_diff.rs:52]: At some point I want to play with a better way of implementing `Packable` (probably just a helper trait) That allows us to cut down the boilerplate a lot. Really, I think that `packed_len`, `unpack` and `pack` should really be possible to derive from the same boilerplate.

Comment (thibault-martinez) [reply bee-ledger/src/types/snapshot/milestone_diff.rs:52]: @pvdrz Started drafting something already and it's in our plans for coordicide to improve it.

File: bee-ledger/src/types/snapshot/kind.rs (hunk 1 above)
Comment (thibault-martinez) [reply bee-ledger/src/types/snapshot/kind.rs:34]: Yeah we'll change Packable soon anyway.

File: bee-ledger/src/types/snapshot/milestone_diff.rs (hunk 2 above)
Comment (thibault-martinez) [reply bee-ledger/src/types/snapshot/milestone_diff.rs]: Done

File: bee-ledger/src/types/snapshot/milestone_diff.rs (hunk 3 above)
Comment (zesterer) [reply bee-ledger/src/types/snapshot/milestone_diff.rs:52]: Aha, nice! I'd like to see that.

**Review by zesterer — APPROVED** [review]

**Review by samuel-rufi — APPROVED** [review]

**Review by Adam-Gleave — APPROVED** [review]

**Pull request merged**